Saule/JsonApiSerializer.cs

```diff
@@ -61,6 +62,7 @@ public PreprocessResult PreprocessContent(object @object, ApiResource resource,
 
                 if (QueryContext != null && !QueryContext.IsHandledQuery)
                 {
+                    var resource = apiResourceProvider.Resolve(dataObject);
```

Comment:
is this needed? in that case, why did the serializer need to be changed?

Reply:
Filtering\Paging\Sorting needs an instance of resource so we need to resolve it based on current `dataObject` and use it for these operations. Before that we had just plain instance of `ApiResource` and after the change we are resolving it on the fly

Reply:
Is the data object never a collection then? I think the serializer now calls `Resolve` for each item in the collection but how does that work with sorting or filtering?

Reply:
correct, it might be collection. and then during the actual serialization it will be called for each item in the collection and might return different `ApiResource` based on the item. In the default implementation it always will be the same `ApiResource` that we have in `ReturnsResource`. 

so sorting\filtering still would work based on default `ApiResource`. if we are talking about our case where we might have different item types in the collection then filterting\sorting would work only for items that exist in base `ApiResource` as it might be very tricky to have different filtering\sorting depending on the item type in the same collection of base types

Reply:
I added 3 more unit tests. Two of them are filtering and sorting by properties in the base `Shape` object and it works. Third one tries to sort by `Left` property in the inherited object `Rectangle` and fails as endpoint returns `IEnumerable<Shape>` and filtering can't find that property in the `Shape` object and throws expected `JsonApiException` with `Attribute 'left' not found.` title

Reply:
Alright that makes sense, if you return a collection of multiple types, you can only sort/filter on properties they share. Nice that you added the tests for that 👍